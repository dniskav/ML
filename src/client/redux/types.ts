import {
  Action,
  PayloadAction,
  TypeConstant,
} from "typesafe-actions";

export interface IReducerAction<TPayload>
  extends Action<TypeConstant>,
  PayloadAction<TypeConstant, TPayload> { }

export interface IApplicationState {
  itemsLoader: boolean;
  items: {
    results: IResultItem[],
    currentItem: IResultItem,
  }
}

export interface IResultItem {
  id: string;
  image: { 
    src: string;
    alt: string;
  };
  price: string;
  freeShipping: boolean;
  currencyId: string;
  data: {
    description: string;
    status: string;
  };
  location: string;
};

export interface IResultCollection {
  results: IResultItem[];
};
