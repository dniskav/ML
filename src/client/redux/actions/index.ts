import { action } from "typesafe-actions";
import axios from 'axios';
export const GET_ITEMS_LIST = "@@actions/GET_ITEMS_LIST";
export const SET_ITEMS_LIST = "@@actions/SET_ITEMS_LIST";
export const SET_ITEM = "@@actions/SET_ITEM";
export const FETCH_ITEMS = "@@actions/FETCH_ITEMS";
export const ITEMS_LOADER = "@@actions/ITEMS_LOADER";


export const setItem = (item: any) => action(SET_ITEM, item);

export const itemsListLoader = (status: boolean) => action(ITEMS_LOADER, status);

export const getitemsList = (items: any) => action(GET_ITEMS_LIST, items);

export const setitemsList = (items: any) => action(SET_ITEMS_LIST, items);

export const fetchItemsList = (query: any) => {
  return async (dispatch: any, getState: any) => {
      dispatch(itemsListLoader(true));
      const state = getState();
      const q = `https://api.mercadolibre.com/sites/MLA/search?q=${query}`;
      try {
          const itemsListRes = await axios.get(q, {});
          dispatch(setitemsList(itemsListRes.data.results));
      } catch (err) {
          console.warn(`ðŸ˜¢ ${err}`, 5);
      } finally {
          dispatch(itemsListLoader(false));
      }
  };
};
