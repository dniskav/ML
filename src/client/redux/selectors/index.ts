import { createSelector } from 'reselect';
import { IApplicationState } from '../types';

const itemSelector = (state: IApplicationState) => state.items;

export const getItem = createSelector(
  itemSelector,
  (state: any) => state,
);