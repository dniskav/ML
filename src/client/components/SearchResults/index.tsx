import React from 'react';
import styled from 'styled-components';
import { IResultCollection } from '../../redux/types';
import ResultItem from '../ResultItem/index';

const StyledSearchResults = styled.div``;

const SearchResults: React.FC<IResultCollection> = ({ results }) => {
  const list = Object.values(results).slice(0, 4);
  return (
    <StyledSearchResults>
      {list.map((item: any) => (
        <ResultItem
          id={item.id}
          key={item?.id}
          freeShipping={item?.shipping?.free_shipping}
          currencyId={item?.currency_id}
          image={{src: item?.thumbnail, alt: item?.title}}
          data={{ description: item?.title, status: item?.condition }}
          price={item?.price}
          location={item?.address?.state_name}
        />
      ))}
    </StyledSearchResults>
  );
};

export default SearchResults;
