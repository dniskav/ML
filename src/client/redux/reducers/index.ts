import { combineReducers } from 'redux';
import { IApplicationState } from '../types';
import itemsReducer from './ItemsReducer';
import { miscReducer } from './miscReducer';

const rootReducer = combineReducers<IApplicationState>({
  items: itemsReducer,
  miscellaneous: miscReducer,
});

export default rootReducer;
