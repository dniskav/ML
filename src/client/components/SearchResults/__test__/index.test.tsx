import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import SearchResults from '../index';
import { IResultCollection } from '../../../redux/types';

const resultCollection: IResultCollection = {
  results: [
    {
      id: "",
      image: { 
        src: "",
        alt: "",
      },
      price: "",
      freeShipping: true,
      currencyId: "",
      data: {
        description: "",
        status: "",
      },
      location: "",
    },
  ],
};

jest.mock('react-router', () => ({
  useParams: () => jest.fn({ id: '1' } as any),
}));

describe('<SearchResults />', () => {
  it('render success', () => {
    const wrapper: ShallowWrapper = shallow(<SearchResults {...resultCollection}/>);
    expect(wrapper).toMatchSnapshot();
  });
});
