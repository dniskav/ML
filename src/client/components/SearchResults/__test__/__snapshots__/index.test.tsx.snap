// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResults /> render success 1`] = `ShallowWrapper {}`;
